### PR TITLE
Added section on discussion about interface approaches

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: courses/advanced-ada/chapters/generics.rst [hunk 1]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
 generic package :ada:`Float_Interface_Pkg.Ops`.
 
-Discussion: Generic interfaces vs. other approaches
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Discussion: formal interfaces vs. other approaches
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:code-config:`reset_accumulator=True`
+
+In Ada, we basically have three approaches to describe interfaces. In
```
Comment (raph-amiard) [inline courses/advanced-ada/chapters/generics.rst]: to describe interfaces *for generic types*. Even though we're in a section on generics I think this needs to be really clear.

File: courses/advanced-ada/chapters/generics.rst [hunk 2]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+addition to the approach using formal interfaces that we've just seen
+above, we also have these approaches:
+
+- Formal subprograms, which we've presented in the introductory course
+  (:doc:`../../intro-to-ada/chapters/generics`).
+
+- Formal packages, which we've discussed in the section on
+  :ref:`formal packages <FormalPackages>`).
```
Comment (raph-amiard) [inline courses/advanced-ada/chapters/generics.rst]: I went to review the section on Formal packages a bit more, and there is some incorrect information here: A signature package is a package with a package without a specification. So the package that you use as an example there is not really a signature package.

The idea of a signature package is that of an interface really. Even though we could expand the terminology, and putting some helpers in the specs is very useful in the real world (akin to default implementations in interfaces), the basic concept should be clarified IMHO.

File: courses/advanced-ada/chapters/generics.rst [hunk 3]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+we first declare the :ada:`My_Type_Hashable` package and use it in the
+declaration of the :ada:`My_Type_Hash_Tables` package.
+
+The advantage of this approach is that the instantiation of the actual
+package (the hash table in our example) is simplified: instead of passing
+all formal subprograms as parameters to :ada:`My_Type_Hash_Tables`, we
+only need to specify the signature package which contains the complete
+interface. When implementing complex interfaces, this approach might lead
+to a cleaner design than the previous approach using formal subprograms
+directly.
```
Comment (raph-amiard) [inline courses/advanced-ada/chapters/generics.rst]: Insisting on the fact that, like in the previous approach, you can have alternative differing implementations might be a good thing !

File: courses/advanced-ada/chapters/generics.rst [hunk 4]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+
+    end Instantiation_Using_Tagged_Types;
 
-.. TODO: Add discussion about interfaces vs. types & formal subprograms
+The instantiation of generic packages is much simpler in this case: we
+don't have to pass operations as parameters in the package instantiation.
+In this example, the declaration of :ada:`My_Type_Hash_Tables` is very
+straightforward: we just have to specify the tagged type (:ada:`My_Type`).
+All operations are *implicitly defined* in the tagged type, so we don't
+have to specify them.
```
Comment (raph-amiard) [inline courses/advanced-ada/chapters/generics.rst]: Conversely, specifying that you're *bound* to use the implementation associated with the type might be interesting here.

File: courses/advanced-ada/chapters/generics.rst [hunk 5]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+          Hash => Alt_Hash);
+
+    end Instantiation_Using_Formal_Function;
+
+Note that, in the declaration of the :ada:`My_Type_Hash_Tables`, we're
+not specifying the :ada:`Hash` function for the instantiation of the
+generic :ada:`Hash_Tables` package. This is possible for two reasons:
+
+- In the declaration of the formal function parameter, we're using
+  :ada:`is <>`, which automatically selects a function with the same name
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: with the same name and a compatible signature

File: courses/advanced-ada/chapters/generics.rst [hunk 6]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+  in the package instantiation if available.
+
+- For :ada:`My_Type`, we've declared a function that has the same name as
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: that has the same name and the expected signature

File: courses/advanced-ada/chapters/generics.rst [hunk 7]
```diff
@@ -1084,19 +1086,335 @@ In this example, we declare the array :ada:`A` with elements of both
...
+form is similar to the approach we've just seen using formal subprograms:
+a signature package defines an interface using a formal type and formal
+subprograms.
+
+Signature packages make it more explicit that the types and subprograms
+defined in the package represent an interface. This is an advantage over
+the approach using formal subprograms directly. However, using signature
+package isn't as explicit as using the :ada:`interface` keyword.
+
+In general, signature packages aren't used in isolation, but in
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: another strong hint is that in general signature packages don't define anything themselves

Comment (raph-amiard) [reply courses/advanced-ada/chapters/generics.rst]: Yep, related to my comment

File: courses/advanced-ada/chapters/generics.rst (hunk 2 above)
Comment (gusthoff) [reply courses/advanced-ada/chapters/generics.rst]: Your comment about signature packages was actually referring to content that was not in this branch. So I rebased this branch, created a separate section to elaborate on signature packages and adapted other parts of the chapter accordingly.

File: courses/advanced-ada/chapters/generics.rst (hunk 4 above)
Comment (gusthoff) [reply courses/advanced-ada/chapters/generics.rst]: Yes, I add this remark to the text. Also, I added a paragraph (include a source-code example) to explain how an alternative implementation can be used.

**Pull request merged**